runtime/arc.js

```diff
@@ -110,15 +110,18 @@ export class Arc {
   }
 
   get idle() {
-    let awaitCompletion = async () => {
-      await this.scheduler.idle;
-      let messageCount = this.pec.messageCount;
-      await this.pec.idle;
-      if (this.pec.messageCount !== messageCount + 1)
-        return awaitCompletion();
-    };
-
-    return awaitCompletion();
+    if (!this._awaitCompletionPromise) {
+      let awaitCompletion = async () => {
+        await this.scheduler.idle;
+        let messageCount = this.pec.messageCount;
+        await this.pec.idle;
+        if (this.pec.messageCount !== messageCount + 1)
+          return awaitCompletion();
+        this._awaitCompletionPromise = null;
```

Comment:
This is *probably* OK but does technically result in _awaitCompletionPromise being set to null all the way back up the stack - should we add an else to prevent that?

Reply:
Can we remove the recursion altogether, say:
```
  get idle() {
    if (!this._awaitCompletionPromise) {
      const awaitCompletion = async () => {
        let messageCount;
        do {
          await this.scheduler.idle;
          messageCount = this.pec.messageCount;
          await this.pec.idle;
        } while (this.pec.messageCount !== messageCount + 1);
        this._awaitCompletionPromise = null;
      };
      this._awaitCompletionPromise = awaitCompletion();
    }
    return this._awaitCompletionPromise;
  }
```

Reply:
I think we can do this and would like to do away with the recursion. I have a tweak to the above and am seeing if I can add a test without too much work. Will update later today.